tests/unit/s2n_serialize_connection_test.c

```diff
@@ -0,0 +1,476 @@
+        };
+
+        /* Serializes TLS 1.3 */
+        if (s2n_is_tls13_fully_supported()) {
+            DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(client_conn);
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(server_conn);
+
+            EXPECT_SUCCESS(s2n_connection_set_config(client_conn, tls13_config));
+            EXPECT_SUCCESS(s2n_connection_set_config(server_conn, tls13_config));
+
+            DEFER_CLEANUP(struct s2n_test_io_stuffer_pair io_pair = { 0 }, s2n_io_stuffer_pair_free);
+            EXPECT_OK(s2n_io_stuffer_pair_init(&io_pair));
+            EXPECT_OK(s2n_connections_set_io_stuffer_pair(client_conn, server_conn, &io_pair));
+
+            EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+            EXPECT_EQUAL(s2n_connection_get_actual_protocol_version(server_conn), S2N_TLS13);
+
+            uint8_t buffer[S2N_SERIALIZED_CONN_TLS13_SHA256_SIZE] = { 0 };
+            EXPECT_SUCCESS(s2n_connection_serialize(server_conn, buffer, sizeof(buffer)));
+
+            struct s2n_blob blob = { 0 };
+            EXPECT_SUCCESS(s2n_blob_init(&blob, buffer, sizeof(buffer)));
+            struct s2n_stuffer stuffer = { 0 };
+            EXPECT_SUCCESS(s2n_stuffer_init_written(&stuffer, &blob));
+
+            uint32_t length = 0;
+            EXPECT_SUCCESS(s2n_connection_serialization_length(server_conn, &length));
+            EXPECT_EQUAL(length, s2n_stuffer_data_available(&stuffer));
+
+            uint64_t serialized_version = 0;
+            EXPECT_SUCCESS(s2n_stuffer_read_uint64(&stuffer, &serialized_version));
+            EXPECT_EQUAL(serialized_version, S2N_SERIALIZED_CONN_V1);
+
+            uint8_t protocol_version[S2N_TLS_PROTOCOL_VERSION_LEN] = { 0 };
+            EXPECT_SUCCESS(s2n_stuffer_read_bytes(&stuffer, protocol_version,
+                    S2N_TLS_PROTOCOL_VERSION_LEN));
+            EXPECT_EQUAL((protocol_version[0] * 10) + protocol_version[1], S2N_TLS13);
+
+            uint8_t cipher_suite[S2N_TLS_CIPHER_SUITE_LEN] = { 0 };
+            EXPECT_SUCCESS(s2n_stuffer_read_bytes(&stuffer, cipher_suite, S2N_TLS_CIPHER_SUITE_LEN));
+            EXPECT_BYTEARRAY_EQUAL(cipher_suite, server_conn->secure->cipher_suite->iana_value,
+                    S2N_TLS_CIPHER_SUITE_LEN);
+
+            uint8_t client_sequence_number[S2N_TLS_SEQUENCE_NUM_LEN] = { 0 };
+            EXPECT_SUCCESS(s2n_stuffer_read_bytes(&stuffer, client_sequence_number,
+                    S2N_TLS_SEQUENCE_NUM_LEN));
+            uint8_t expected_sequence_number[] = { 0, 0, 0, 0, 0, 0, 0, 0 };
+            EXPECT_BYTEARRAY_EQUAL(client_sequence_number, expected_sequence_number,
+                    S2N_TLS_SEQUENCE_NUM_LEN);
```

Comment:
Why is the sequence number 0? Isn't it updated when sending records during the handshake?

Reply:
Yeah this is an interesting difference between TLS1.3 and TLS1.2. The TLS1.2 application keys get used to encrypt/decrypt the last message in the TLS1.2 handshake. However in TLS1.3, those messages are encrypted with the handshake keys, not the traffic keys. So post-handshake the sequence numbers for TLS1.3 are 0, and 1 in TLS1.2.

Reply:
ahh got it okay, interesting. thanks!